### PR TITLE
Add renameat2

### PR DESCRIPTION
**Pull request by cyberhuman**
- add renameat2 function and flags
- suffix constructors marked with `[@have <feature>]` with `__Not_available` if  <feature> is not available.
- register `ExtUnix.Not_available` with C runtime so it's possible to raise `Not_available` from bindings

### CODE REVIEW

File: src/rename.c [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+CAMLprim value caml_extunix_renameat2(value v_oldfd, value v_oldname, value v_newfd, value v_newname, value v_flags)
+{
+  CAMLparam5(v_oldfd, v_oldname, v_newfd, v_newname, v_flags);
+  check_flag_list(v_flags);
+  int flags = caml_convert_flag_list(v_flags, rename_flags_table);
+  caml_enter_blocking_section();
+  int ret = renameat2(Int_val(v_oldfd), String_val(v_oldname), Int_val(v_newfd), String_val(v_newname), flags);
+  caml_leave_blocking_section();
```
Comment (MisterDA) [inline src/rename.c:36]: This is incorrect, it is not allowed to use `X_val` macros inside a blocking section.

File: discover/discover.ml [hunk 2]
```diff
@@ -229,6 +230,16 @@ let features =
       D "S_IFREG";
       S "fstatat"; S "openat"; S "unlinkat"; S "renameat"; S "mkdirat"; S "linkat"; S "symlinkat"; S "readlinkat"; S "fchownat"; S "fchmodat";
     ];
+    "RENAMEAT2", L[
+      fd_int;
+      DEFINE "_GNU_SOURCE";
```
Comment (ygrek) [inline discover/discover.ml:235]: technically not needed, _GNU_SOURCE is defined for all, see config_defines

File: src/rename.c [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+CAMLprim value caml_extunix_renameat2(value v_oldfd, value v_oldname, value v_newfd, value v_newname, value v_flags)
+{
+  CAMLparam5(v_oldfd, v_oldname, v_newfd, v_newname, v_flags);
+  check_flag_list(v_flags);
+  int oldfd = Int_val(v_oldfd), newfd = Int_val(v_newfd);
+  const char *oldname = String_val(v_oldname), *newname = String_val(v_newname);
+  int flags = caml_convert_flag_list(v_flags, rename_flags_table);
+  caml_enter_blocking_section();
+  int ret = renameat2(oldfd, oldname, newfd, newname, flags);
```
Comment (ygrek) [inline src/rename.c:37]: this is not safe, oldname and newname still refer to ocaml heap, need to strdup (or remove blocking section)

**Review by ygrek — CHANGES_REQUESTED** [review]
sorry, didn't see this on time
nice idea with not available constructors

**Comment by ygrek** [comment]
closed in favor of #57

**Pull request closed without merging**